From 5ddb78bce2bd6ef71c2c7bcbdedfcd0f577ec740 Mon Sep 17 00:00:00 2001
From: Pierre-Hugues Husson <phh@phh.me>
Date: Sun, 18 Dec 2022 18:17:30 -0500
Subject: [PATCH] FOD support for Samsung and Asus

[los23-port] rebased: BLASTBufferQueue consumer creation moved ctor->initialize() upstream (uses mName); OutputLayer/ComposerHal/HWC2 context adjusted; vendor800_1 inherited from V2_1 CommandWriterBase (hardware_interfaces patch)

Change-Id: Ic142a5ff7ddc85cb8d1c98bad584932c67488b0d
---
 libs/gui/BLASTBufferQueue.cpp                 | 20 ++++++++++++++--
 libs/ui/Gralloc2.cpp                          |  1 -
 libs/ui/Gralloc3.cpp                          |  1 -
 libs/ui/Gralloc4.cpp                          |  1 -
 .../CompositionEngine/src/OutputLayer.cpp     | 24 +++++++++++++++++++
 .../DisplayHardware/AidlComposerHal.cpp       |  8 +++++++
 .../DisplayHardware/AidlComposerHal.h         |  3 +++
 .../DisplayHardware/ComposerHal.h             |  3 +++
 .../surfaceflinger/DisplayHardware/HWC2.cpp   |  8 +++++++
 .../surfaceflinger/DisplayHardware/HWC2.h     |  6 +++++
 .../DisplayHardware/HidlComposerHal.cpp       |  7 ++++++
 .../DisplayHardware/HidlComposerHal.h         |  3 +++
 12 files changed, 80 insertions(+), 5 deletions(-)

diff --git a/libs/gui/BLASTBufferQueue.cpp b/libs/gui/BLASTBufferQueue.cpp
index 05fbc148d2..75792c412b 100644
--- a/libs/gui/BLASTBufferQueue.cpp
+++ b/libs/gui/BLASTBufferQueue.cpp
@@ -41,6 +41,7 @@
 
 #include <private/gui/ComposerService.h>
 #include <private/gui/ComposerServiceAIDL.h>
+#include <cutils/properties.h>
 
 #include <android-base/thread_annotations.h>
 
@@ -49,6 +50,14 @@
 using namespace com::android::graphics::libgui;
 using namespace std::chrono_literals;
 
+static bool sCheckedProps = false;
+static bool sSamsungFod = false;
+static void init_fod_props() {
+    if(sCheckedProps) return;
+    sCheckedProps = true;
+    sSamsungFod = property_get_bool("persist.sys.phh.fod.samsung", false);
+}
+
 namespace {
 
 template <class Mutex>
@@ -194,10 +203,17 @@ void BLASTBufferItemConsumer::resizeFrameEventHistory(size_t newSize) {
 void BLASTBufferQueue::initialize() {
     std::lock_guard _lock{mMutex};
     createBufferQueue(&mProducer, &mConsumer);
+    uint64_t usage = GraphicBuffer::USAGE_HW_COMPOSER |
+        GraphicBuffer::USAGE_HW_TEXTURE;
+
+    init_fod_props();
+    if(sSamsungFod && mName.find("SurfaceView[UdfpsControllerOverlay]") != std::string::npos) {
+           usage |= 0x400000000LL;
+    }
+
     mBufferItemConsumer =
             sp<BLASTBufferItemConsumer>::make(mProducer, mConsumer,
-                                              GraphicBuffer::USAGE_HW_COMPOSER |
-                                                      GraphicBuffer::USAGE_HW_TEXTURE,
+                                              usage,
                                               1, false, wp<BLASTBufferQueue>::fromExisting(this));
     // since the adapter is in the client process, set dequeue timeout
     // explicitly so that dequeueBuffer will block
diff --git a/libs/ui/Gralloc2.cpp b/libs/ui/Gralloc2.cpp
index 41306b51f6..3badc79606 100644
--- a/libs/ui/Gralloc2.cpp
+++ b/libs/ui/Gralloc2.cpp
@@ -115,7 +115,6 @@ status_t Gralloc2Mapper::validateBufferDescriptorInfo(
     if (descriptorInfo->usage & ~validUsageBits) {
         ALOGE("buffer descriptor contains invalid usage bits 0x%" PRIx64,
               descriptorInfo->usage & ~validUsageBits);
-        return BAD_VALUE;
     }
 
     // Gralloc2 implementations never understand non-BLOB with GPU_DATA_BUFFER
diff --git a/libs/ui/Gralloc3.cpp b/libs/ui/Gralloc3.cpp
index e9aa9c8d54..f29ce686c2 100644
--- a/libs/ui/Gralloc3.cpp
+++ b/libs/ui/Gralloc3.cpp
@@ -106,7 +106,6 @@ status_t Gralloc3Mapper::validateBufferDescriptorInfo(
     if (descriptorInfo->usage & ~validUsageBits) {
         ALOGE("buffer descriptor contains invalid usage bits 0x%" PRIx64,
               descriptorInfo->usage & ~validUsageBits);
-        return BAD_VALUE;
     }
 
     // Gralloc3 implementations never understand non-BLOB with GPU_DATA_BUFFER
diff --git a/libs/ui/Gralloc4.cpp b/libs/ui/Gralloc4.cpp
index f1322f1e4c..c3ff48a262 100644
--- a/libs/ui/Gralloc4.cpp
+++ b/libs/ui/Gralloc4.cpp
@@ -129,7 +129,6 @@ static status_t validateBufferDescriptorInfo(IMapper::BufferDescriptorInfo* desc
     if (descriptorInfo->usage & ~validUsageBits) {
         ALOGE("buffer descriptor contains invalid usage bits 0x%" PRIx64,
               descriptorInfo->usage & ~validUsageBits);
-        return BAD_VALUE;
     }
 
     // Combinations that are only allowed with gralloc 4.1.
diff --git a/services/surfaceflinger/CompositionEngine/src/OutputLayer.cpp b/services/surfaceflinger/CompositionEngine/src/OutputLayer.cpp
index 36a9df9ded..3652e3e642 100644
--- a/services/surfaceflinger/CompositionEngine/src/OutputLayer.cpp
+++ b/services/surfaceflinger/CompositionEngine/src/OutputLayer.cpp
@@ -25,6 +25,7 @@
 #include <compositionengine/impl/OutputLayerCompositionState.h>
 #include <ui/FloatRect.h>
 #include <ui/HdrRenderTypeUtils.h>
+#include <cutils/properties.h>
 #include <cstdint>
 #include <limits>
 #include "system/graphics-base-v1.0.h"
@@ -40,6 +41,9 @@
 // TODO(b/129481165): remove the #pragma below and fix conversion issues
 #pragma clang diagnostic pop // ignored "-Wconversion"
 
+static bool sCheckedProps = false;
+static bool sAsusFod = false;
+
 using aidl::android::hardware::graphics::composer3::Composition;
 using aidl::android::hardware::graphics::composer3::Luts;
 
@@ -539,6 +543,26 @@ void OutputLayer::writeOutputDependentGeometryStateToHWC(HWC2::Layer* hwcLayer,
                                                      ->getHeight()));
     }
 
+    if(!sCheckedProps) {
+        sCheckedProps = true;
+        sAsusFod = property_get_bool("persist.sys.phh.fod.asus", false);
+    }
+
+    if (strstr(getLayerFE().getDebugName(), "UdfpsControllerOverlay#") != nullptr) {
+        if (sAsusFod) {
+            if (auto error = hwcLayer->setLayerClass(5); error != hal::Error::NONE) {
+                ALOGE("Failed setting Asus layer class");
+            }
+        }
+    }
+    if (strstr(getLayerFE().getDebugName(), "SurfaceView[UdfpsControllerOverlay](BLAST)#") != nullptr) {
+        if (sAsusFod) {
+            if (auto error = hwcLayer->setLayerClass(4); error != hal::Error::NONE) {
+                ALOGE("Failed setting Asus layer class");
+            }
+        }
+    }
+
     ALOGV("Writing display frame [%d, %d, %d, %d]", displayFrame.left, displayFrame.top,
           displayFrame.right, displayFrame.bottom);
 
diff --git a/services/surfaceflinger/DisplayHardware/AidlComposerHal.cpp b/services/surfaceflinger/DisplayHardware/AidlComposerHal.cpp
index 0ab9a4c597..d7b608c069 100644
--- a/services/surfaceflinger/DisplayHardware/AidlComposerHal.cpp
+++ b/services/surfaceflinger/DisplayHardware/AidlComposerHal.cpp
@@ -1883,5 +1883,13 @@ void AidlComposer::addDisplay(Display display) {
 void AidlComposer::onHotplugConnect(Display display) {
     addDisplay(display);
 }
+
+Error AidlComposer::setLayerClass(Display display, Layer layer, uint32_t layerClass) {
+    (void) display;
+    (void) layer;
+    (void) layerClass;
+    return Error::NONE;
+}
+
 } // namespace Hwc2
 } // namespace android
diff --git a/services/surfaceflinger/DisplayHardware/AidlComposerHal.h b/services/surfaceflinger/DisplayHardware/AidlComposerHal.h
index 818ca0c5e2..680bed1d4c 100644
--- a/services/surfaceflinger/DisplayHardware/AidlComposerHal.h
+++ b/services/surfaceflinger/DisplayHardware/AidlComposerHal.h
@@ -244,6 +244,9 @@ public:
     Error setRefreshRateChangedCallbackDebugEnabled(Display, bool) override;
     Error notifyExpectedPresent(Display, nsecs_t expectedPresentTime,
                                 int32_t frameIntervalNs) override;
+
+    // Proprietary extensions
+    Error setLayerClass(Display display, Layer layer, uint32_t layerClass) override;
     Error getRequestedLuts(
             Display display, std::vector<Layer>* outLayers,
             std::vector<aidl::android::hardware::graphics::composer3::DisplayLuts::LayerLut>*
diff --git a/services/surfaceflinger/DisplayHardware/ComposerHal.h b/services/surfaceflinger/DisplayHardware/ComposerHal.h
index 76cbcb6049..0e0c1454bb 100644
--- a/services/surfaceflinger/DisplayHardware/ComposerHal.h
+++ b/services/surfaceflinger/DisplayHardware/ComposerHal.h
@@ -315,6 +315,9 @@ public:
     virtual Error setRefreshRateChangedCallbackDebugEnabled(Display, bool) = 0;
     virtual Error notifyExpectedPresent(Display, nsecs_t expectedPresentTime,
                                         int32_t frameIntervalNs) = 0;
+
+    // Proprietary extensions
+    virtual Error setLayerClass(Display display, Layer layer, uint32_t layerClass) = 0;
     virtual Error getRequestedLuts(Display display, std::vector<Layer>* outLayers,
                                    std::vector<V3_0::DisplayLuts::LayerLut>* outLuts) = 0;
     virtual Error setLayerLuts(Display display, Layer layer, V3_0::Luts& luts) = 0;
diff --git a/services/surfaceflinger/DisplayHardware/HWC2.cpp b/services/surfaceflinger/DisplayHardware/HWC2.cpp
index ecaf888765..3468f29a0c 100644
--- a/services/surfaceflinger/DisplayHardware/HWC2.cpp
+++ b/services/surfaceflinger/DisplayHardware/HWC2.cpp
@@ -1149,6 +1149,14 @@ Error Layer::setPictureProfileHandle(const PictureProfileHandle& handle) {
     return static_cast<Error>(intError);
 }
 
+Error Layer::setLayerClass(uint32_t layerClass) {
+    if (CC_UNLIKELY(!mDisplay)) {
+        return Error::BAD_DISPLAY;
+    }
+    auto intError = mComposer.setLayerClass(mDisplay->getId(), mId, layerClass);
+    return static_cast<Error>(intError);
+}
+
 } // namespace impl
 } // namespace HWC2
 } // namespace android
diff --git a/services/surfaceflinger/DisplayHardware/HWC2.h b/services/surfaceflinger/DisplayHardware/HWC2.h
index 5b2c6446c9..a789407a83 100644
--- a/services/surfaceflinger/DisplayHardware/HWC2.h
+++ b/services/surfaceflinger/DisplayHardware/HWC2.h
@@ -404,6 +404,9 @@ public:
     // AIDL HAL
     [[nodiscard]] virtual hal::Error setBrightness(float brightness) = 0;
     [[nodiscard]] virtual hal::Error setBlockingRegion(const android::Region& region) = 0;
+
+    // Proprietary HAL
+    [[nodiscard]] virtual hal::Error setLayerClass(uint32_t layerClass) = 0;
     [[nodiscard]] virtual hal::Error setLuts(
             aidl::android::hardware::graphics::composer3::Luts& luts) = 0;
     [[nodiscard]] virtual hal::Error setPictureProfileHandle(
@@ -458,6 +461,9 @@ public:
     // AIDL HAL
     hal::Error setBrightness(float brightness) override;
     hal::Error setBlockingRegion(const android::Region& region) override;
+
+    // Proprietary HAL
+    hal::Error setLayerClass(uint32_t layerClass) override;
     hal::Error setLuts(aidl::android::hardware::graphics::composer3::Luts&) override;
     hal::Error setPictureProfileHandle(const PictureProfileHandle& handle) override;
 
diff --git a/services/surfaceflinger/DisplayHardware/HidlComposerHal.cpp b/services/surfaceflinger/DisplayHardware/HidlComposerHal.cpp
index 0bf5ef371f..a84c116e69 100644
--- a/services/surfaceflinger/DisplayHardware/HidlComposerHal.cpp
+++ b/services/surfaceflinger/DisplayHardware/HidlComposerHal.cpp
@@ -1490,6 +1490,13 @@ Error HidlComposer::setLayerPictureProfileId(Display, Layer, PictureProfileId) {
     return Error::UNSUPPORTED;
 }
 
+Error HidlComposer::setLayerClass(Display display, Layer layer, uint32_t layerClass) {
+    mWriter.selectDisplay(display);
+    mWriter.selectLayer(layer);
+    mWriter.vendor800_1(layerClass);
+    return Error::NONE;
+}
+
 void HidlComposer::registerCallback(ComposerCallback& callback) {
     const bool vsyncSwitchingSupported =
             isSupported(Hwc2::Composer::OptionalFeature::RefreshRateSwitching);
diff --git a/services/surfaceflinger/DisplayHardware/HidlComposerHal.h b/services/surfaceflinger/DisplayHardware/HidlComposerHal.h
index 2c885dc22e..ff536e7212 100644
--- a/services/surfaceflinger/DisplayHardware/HidlComposerHal.h
+++ b/services/surfaceflinger/DisplayHardware/HidlComposerHal.h
@@ -355,6 +355,9 @@ public:
                                    Hdr*) override;
     Error setRefreshRateChangedCallbackDebugEnabled(Display, bool) override;
     Error notifyExpectedPresent(Display, nsecs_t, int32_t) override;
+
+    // Proprietary extensions
+    Error setLayerClass(Display display, Layer layer, uint32_t layerClass) override;
     Error getRequestedLuts(
             Display, std::vector<Layer>*,
             std::vector<aidl::android::hardware::graphics::composer3::DisplayLuts::LayerLut>*)
-- 
2.43.0

